staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go

```diff
@@ -263,10 +264,15 @@ func (f *ConfigFlags) toDiscoveryClient() (discovery.CachedDiscoveryInterface, e
 		return nil, err
 	}
 
-	// The more groups you have, the more discovery requests you need to make.
-	// given 25 groups (our groups + a few custom resources) with one-ish version each, discovery needs to make 50 requests
-	// double it just so we don't end up here again for a while.  This config is only used for discovery.
-	config.Burst = f.discoveryBurst
+	// We don't rate limit discovery by default, preferring to defer to APF.
+	// https://github.com/kubernetes/enhancements/tree/11a976c/keps/sig-api-machinery/1040-priority-and-fairness
+	config.RateLimiter = flowcontrol.NewFakeAlwaysRateLimiter()
```

Comment:
This rate limiter does what I want (i.e. disables rate limiting) but I feel a bit uncomfortable using it given it seems to be intended as a test fake.

I wonder if it would make more sense to make this change (i.e. disabling rate limiting for discovery) down in client-go's discovery client defaults. Doing so would presumably have a much larger blast radius.

https://github.com/kubernetes/client-go/blob/fd0a0345/discovery/discovery_client.go#L469

Reply:
Is there a chance that disabling the rate limiter for discovery is breaking API servers? It might be fine for 2,000 CRDs, but what about 20,000? Or 200,000? If there's no protection built in that seems dangerous.

It almost seems like discovery has to be done more efficiently by client-go and not query every single CRD every 10 minutes.
I do agree that it probably should be done in client-go though, because the slow down also happens for controllers (not just kubectl)

Reply:
> Is there a chance that disabling the rate limiter for discovery is breaking API servers? It might be fine for 2,000 CRDs, but what about 20,000? Or 200,000? If there's no protection built in that seems dangerous.

See https://github.com/kubernetes/enhancements/tree/11a976c/keps/sig-api-machinery/1040-priority-and-fairness for server-side protection.